### ISSUE TITLE
ServiceClosureArgument is not recognized by lint:container

### ISSUE THREAD

**Issue by vudaltsov** [symfony/symfony#39259]
**Symfony version(s) affected**: 5.2, 5.1, probably earlier versions too

**Description**  
When service is injected via `ServiceClosureArgument` like in [DoctrineMigrationsExtension](https://github.com/doctrine/DoctrineMigrationsBundle/blob/a8b0114b9672a44d7e045c07df9578d72d695978/DependencyInjection/DoctrineMigrationsExtension.php#L74), `lint:container` throws an error:

> [ERROR] Invalid definition for service "doctrine.migrations.dependency_factory": argument 2 of "Doctrine\Migrations\DependencyFactory::setDefinition" accepts "callable", "App\CustomSchemaProvider" passed.

**How to reproduce**  
https://github.com/vudaltsov/symfony-lint-container-reproducer

**UPDATE**
Since `doctrine/doctrine-migrations-bundle` v3.1.0 container lint fails without any custom service configuration.

**Comment by vudaltsov** [comment]
@nicolas-grekas , need your help with this. I investigated the issue a bit and it looks strange.

First, there is already an `if` to support `ServiceClosureArgument`, see [line](https://github.com/symfony/symfony/blob/2d062df845dd40d56f2ca9a483fc3f01e6b49a0a/src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php#L242). And it's covered with a non-failing test. Good, but it does not work in the real project!

The problem is that at the time the container builder is used in `ContainerLintCommand` ([this line](https://github.com/symfony/symfony/blob/2d062df845dd40d56f2ca9a483fc3f01e6b49a0a/src/Symfony/Bundle/FrameworkBundle/Command/ContainerLintCommand.php#L124)), the method call is not a `ServiceClosureArgument` anymore, it's a reference, so the reference's class is checked against `callable`, not `ServiceClosureArgument`:

![image](https://user-images.githubusercontent.com/2552865/101018904-34ec0800-357d-11eb-98fe-1a1a312c3abf.png)

At the same time in the dumped container it's a closure, which is correct:

![image](https://user-images.githubusercontent.com/2552865/101019113-87c5bf80-357d-11eb-9d88-bf1a95a49a7e.png)

**Comment by stof** [comment]
> The problem is that at the time the container builder is used in `ContainerLintCommand` ([this line](https://github.com/symfony/symfony/blob/2d062df845dd40d56f2ca9a483fc3f01e6b49a0a/src/Symfony/Bundle/FrameworkBundle/Command/ContainerLintCommand.php#L124)), the method call is not a `ServiceClosureArgument` anymore, it's a reference

that's what need to be investigated. I don't see any valid reason for it to be a Reference object here.

**Comment by vudaltsov** [comment]
I found the issue. `ContainerLintCommand` uses xml dump to recreate `ContainerBuilder`, but `XmlDumper` dumps `ServiceClosureArgument` as a value, not as a value wrapped in a closure:

https://github.com/symfony/symfony/blob/31ee43a8c67897a2f61782060965f5c8a81f58a2/src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php#L279

Here's my idea. To add a static method `ServiceClosureArgument::wrapServiceInAClosure()` to be able to use it as a factory for an inline service:

```php
final class ServiceClosureArgument
{
    // ...

    public static function wrapServiceInAClosure($service): \Closure
    {
        return function () use ($service) {
            return $service;
        };
    }
}
```

```xml
<call method="setDefinition">
    <argument>Doctrine\Migrations\Provider\SchemaProvider</argument>
    <argument type="service">
        <service class="Closure">
            <factory class="Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument" 
                     method="wrapServiceInAClosure"/>
            <argument type="service" id="HappyInc\Doctrine\SchemaConfigurator\CompositeSchemaProvider"/>
        </service>
    </argument>
</call>
```

**Comment by stof** [comment]
yeah, I think the issue is that we don't have a XML syntax to create a ServiceClosureArgument right now. they can only be created from PHP config. That's what we should fix instead IMO, rather than adding a factory method

**Comment by crmpicco** [comment]
@vudaltsov Nice find. I'm curious to know if there has been any progress with an upstream fix for this?

**Comment by derrabus** [comment]
Thanks for the ping @crmpicco. I wasn't aware of this issue and opened another one in the meantime: #40773

**Comment by mbrodala** [comment]
Notice: the same issue also occurs with `doctrine/doctrine-migrations-bundle` 3.0.x already.

Workaround for now:

```
$ bin/console lint:container --env prod
                                                                                                                        
 [OK] The container was lint successfully: all services are injected with values that are compatible with their type    
      declarations.                                                                                                     
                                                                                                                        
```

**Comment by nicolas-grekas** [comment]
Will be fixed by #41176
